### DISCUSSION TITLE
Update custom provider examples in source to KC 17.x

### DISCUSSION THREAD

**Discussion by thaDude** [keycloak/keycloak#10258] [Archive]
Hello,

Are there plans to update the custom REST endpoint examples in https://github.com/keycloak/keycloak/tree/17.0.0/examples/providers to Keycloak 17 based on Quarkus?

I have only recently finished the effort of coding my own endpoints and - somewhat painstakingly - scripting Wildfly to deploy these extensions as part of a Docker build.

Best regards,

David

**Comment by pedroigor** [comment]
@thaDude Created an issue for it.

The steps are less painful as we no longer use XML as a configuration format.

In this case, you only need to copy the jar to te `providers` directory. And the run:

```
kc.sh build <OPTIONS>
```

And then start the server with the configuration options you need. 

You can also skip the `build` step and just run `kc start --auto-build` right after copying the jar.

Or in development mode, just run `start-dev`.

**Reply by mteubner** [reply]
Currently I stopped the investigation of 17.0 because none of our prior used self-written providers are currently working in the new deployment style. Mainly with the same issue like https://github.com/keycloak/keycloak/issues/10230.

**Reply by pedroigor** [reply]
@mteubner Could you please check if this new guide helps you? See https://github.com/keycloak/keycloak/pull/10401.

**Reply by mteubner** [reply]
Not quite, sorry if I misunderstand, and thank you for try to clear it up for me.

I added this property to the keycloak.conf in conf folder and when I run kc.sh show-config it is displayed.

kc.spi-client-jpa.searchableAttributes =  [tenant_id] (PropertiesConfigSource[source=file:/opt/keycloak/bin/../conf/keycloak.conf])

But debugging the start up, I can see that JpaClientProviderFactory still don't read it also it is in the Properties sources with the right prefix.

![image](https://user-images.githubusercontent.com/58938193/155326340-90348ac6-b83b-4626-b734-f7223553b92a.png)

**Reply by mteubner** [reply]
Of course I can still do the -e JAVA_OPTS_APPEND="-Dkeycloak.client.searchableAttributes=first_attribute,second_attribute" but it would be nicer getting it through with the keycloak.conf

**Reply by mteubner** [reply]
I FOUND IT :)

So the camel case property name will be split by hyphen too. So the right property to set is

spi-client-jpa-searchable-attributes

**Reply by mteubner** [reply]
And as environment variable it would be 

KC_SPI_CLIENT_JPA_SEARCHABLE_ATTRIBUTES

**Reply by mteubner** [reply]
Thanks a lot

**Comment by mteubner** [comment]
Also I still have the problem to just use my old providers, when I have some time I will try examples\providers\authenticator\ if it has the same problem. So still examples would be nice.

I have a Jar with two providers with factories but one is not going to be accepted anymore by keycloak on build.

The factory is still configured in META-INF/services. It passes the session and the scope related searchable attributes to the provider.

`@Override
  public RealmResourceProvider create(KeycloakSession session) {
    return new ClientResourceProvider(session, clientSearchableAttributes);
  }

  @Override
  public void init(Scope config) {
    String[] searchableAttrsArr = config.getArray("searchableAttributes");
    if (searchableAttrsArr == null) {
      String s = System.getProperty("keycloak.client.searchableAttributes");
      searchableAttrsArr = s == null ? null : s.split("\\s*,\\s*");
    }
    HashSet<String> s = new HashSet<>(REQUIRED_SEARCHABLE_ATTRIBUTES);
    if (searchableAttrsArr != null) {
      s.addAll(Arrays.asList(searchableAttrsArr));
    }
    clientSearchableAttributes = Collections.unmodifiableSet(s);
  }`

But when the Jar is deployed and bin/kc.sh build is called, the following exceptions is raised.
`
    ERROR: Found 3 deployment problems:
[1] Unsatisfied dependency for type org.keycloak.models.KeycloakSession and qualifiers [@Default]
    - java member: com.gk_software.cloud.keycloak.provider.rest.ClientResourceProvider().session
        - declared on CLASS bean [types=[ClientResourceProvider, org.keycloak.provider.Provider, org.keycloak.services.resource.RealmResourceProvider, AdminAuthResource, java.lang.Object, CustomClientEndpoint], qualifiers=[@Default, @Any], target=ClientResourceProvider]

[2] Unsatisfied dependency for type java.util.Set<java.lang.String> and qualifiers [@Default]
    - java member: com.gk_software.cloud.keycloak.provider.rest.ClientResourceProvider().clientSearchableAttributes
        - declared on CLASS bean [types=[ClientResourceProvider, org.keycloak.provider.Provider, org.keycloak.services.resource.RealmResourceProvider,AdminAuthResource, java.lang.Object, CustomClientEndpoint], qualifiers=[@Default, @Any], target=ClientResourceProvider]
    The following beans match by type, but none have matching qualifiers:
    - Bean [class=java.util.Set, qualifiers=[@ConfigProperty, @Any]]
`

I currently struggling to understand what actually is missing for the injection.

**Reply by mteubner** [reply]
When I remove any constructor parameter the providers work again, but strange is that the HelloExample is working also with session as constructor parameter.
